lib/pedicel/base.rb

```diff
@@ -140,58 +155,93 @@ def verify_signature(ca_certificate_pem: Pedicel.config[:apple_root_ca_g3_cert_p
       self
     end
 
-    private
+    def self.extract_certificates(signature:, config: Pedicel.config)
+      leafs, intermediates, others = [], [], []
 
-    def self.verify_signature_certificate_oids(signature:)
-      leaf = signature.certificates.find{|c| c.extensions.find{|e| e.oid == Pedicel.config[:oids][:leaf_certificate]}}
-      unless leaf
-        raise SignatureError, "no leaf certificate found (OID #{Pedicel.config[:oids][:leaf_certificate]})"
-      end
+      signature.certificates.each do |certificate|
+        leaf_or_intermediate = false
+
+        certificate.extensions.each do |extension|
+          case extension.oid
+          when config[:oids][:intermediate_certificate]
+            intermediates << certificate
+            leaf_or_intermediate = true
+            break
+          when config[:oids][:leaf_certificate]
+            leafs << certificate
+            leaf_or_intermediate = true
+            break
+          end
```

Comment:
A certificate can have multiple extensions. But ... If a certificate has both the intermediate and leaf OID ... then ... `break` is wrong. Since the cert should then be added to both `intermediates` and `leafs` (which will most likely result in an error).

This is a nice catch!

Reply:
I'll add a (pending) spec for this case.